src/main/java/hudson/tasks/junit/TestResult.java

```diff
@@ -222,6 +256,98 @@ public TestResult getTestResult() {
     	return this;
     }
 
+    public TestResult(TestResult src) {
+        this.stdioRetention = src.stdioRetention;
+        this.keepProperties = src.keepProperties;
+        this.keepTestNames = src.keepTestNames;
+        this.duration = src.duration;
+        if (src.suites != null) {
+            this.suites = new ArrayList<SuiteResult>();
+            for (SuiteResult sr : src.suites) {
+                suites.add(new SuiteResult(sr));
+            }
+        } else {
+            this.suites = null;
+        }
+        this.impl = null;
+    }
+
+    private static final XMLInputFactory factory = XMLInputFactory.newInstance();
+    public void parse(XmlFile f) throws XMLStreamException, IOException {
+        try (Reader r = f.readRaw()){
+            final XMLEventReader reader = factory.createXMLEventReader(r);
+            while (reader.hasNext()) {
+                final XMLEvent event = reader.nextEvent();
+                if (event.isStartElement() && event.asStartElement().getName()
+                        .getLocalPart().equals("result")) {
+                    parseXmlResult(reader, event.asStartElement());
+                }
+            }
+            r.close();
+        }
+    }
```

Comment:
does this not create an insecure Parser?

The default may or may not support external entities leading to SSRF/DOS etc

Reply:
To me this is slightly confusing, as the parser is technically the **parse** function, I would assume that such features require support within this function to have any effect.

I ran some examples from that webpage through a Jenkins instance and nothing happened, maybe someone could shed more light on it?

Reply:
Can this example be adapted to use /dev/random ? I read it should hang the server? I have never used XML like this before.

Reply:
> Can this example be adapted to use /dev/random

you should be able to use any file that you can read.  I thought (maybe incorrectly , it's been a while) that this in Java would show a directory listing.

<!DOCTYPE result[
  <!ENTITY foo SYSTEM "file:///dev/random">
]>

if you are in a unit test then you can use the URL of a resource in src/test/resources to be cross platform.

Reply:
Does it make a difference if I replace XMLEventReader with XMLStreamReader. It is a drop-in replacement, basically. Based on the docs, it should be doing even less processing on its own.

Reply:
> Does it make a difference if I replace XMLEventReader with XMLStreamReader. It is a drop-in replacement, basically. Based on the docs, it should be doing even less processing on its own.

It _may_ do, but I am not 100% confident as the security settings are set on the `XMLInputFactory` not the the reader, you will have to do more work to process the stream.

Do you need to support DTD or external or even entities here?
if not then the following should (untested) work and produce a safe parser.

```java
private static XMLInputFactory createFactory() {
    XMLInputFactory factory = XMLInputFactory.newInstance();
    factory.setProperty(XMLInputFactory.SUPPORT_DTD, Boolean.FALSE);
    factory.setProperty(XMLInputFactory.IS_REPLACING_ENTITY_REFERENCES, Boolean.FALSE);
    return factory;
}
```

you can take a look at https://github.com/openjdk/jdk/blob/8feabc849ba2f617c8c6dbb2ec5074297beb6437/test/jaxp/javax/xml/jaxp/unittest/stream/XMLStreamReaderTest/SupportDTDTest.java to see what this does to the code and experiment.

Reply:
Note, if you are worried about regressions due to these settings you could use an escape hatch like 
https://github.com/jenkinsci/jenkins/blob/ef376fe6f61731ae2acee82a2ea8a760003f7b21/core/src/main/java/jenkins/util/xml/XMLUtils.java#L45 to not set these properties.  (and then if there are no reports of issues the escape hatch can be removed in a future version)

Reply:
Switched to XMLStreamReader and also set the properties, and added an escape hatch.